ETag middleware with implementation for OpenAPI

lzap opened the pull request:
This new middleware can be used to cache any static or nearly static content. Everything works transparently, as long as a valid etag (string, hash, datetime, timestamp, anything unique) is provided and expiration time is set.

Caching works transparently in browsers, it will not work when API is hit from a curl utility of course.

The OpenAPI caching is just an example, this will be useful for instance types endpoint once we start embedding them into the application instead of dynamically loading them every time.

It can be also useful for any content that was cached in the future.

Thanks for the contribution, take a moment to read our contributing and security guidelines:

I tested it works fine with JSON spec, however, I am trying to open up redoc documentation and it simply renders JSON instead of giving me any HTML. What is the URL again?

http://localhost:8000/docs/openapi.json

lzap commented:
Rebased not to use sha1 which is crypto-weak and liner complains about it. It now uses crc64 which is okay for etag.

lzap commented on api/embed.go:
```diff
@@ -3,12 +3,22 @@ package api
+	hash.Write(embeddedJSONSpec)
+	ETagJSONSpec = fmt.Sprintf("%x", hash.Sum64())
+}
+
```
Calculates CRC64 hash of the whole JSON spec file during startup and that is returned to the browser.

lzap commented on internal/middleware/etag_middleware.go:
```diff
@@ -0,0 +1,39 @@
+		}
+		return http.HandlerFunc(fn)
+	}
+}
```
The middleware responsible for setting the ETag and Cache-Control and respecting the If-None-Match header from the client to quickly return NotModified if browser already has that specific version.

lzap commented on internal/routes/all_routes.go:
```diff
@@ -23,15 +24,21 @@ func SetupRoutes(r *chi.Mux) {
+	r.Route("/openapi.json", func(r chi.Router) {
+		r.Use(middleware.ETagMiddleware(15*time.Minute, "json-spec", api.ETagJSONSpec))
+		r.Get("/", api.ServeOpenAPISpec)
+	})
```
And this is how to use the middleware, I set the timeout to 15 minutes. I use it both for JSON spec and documentation.

It's not super useful for OpenAPI, granted, but it's a good example. A good candidate will definitely be instance-types endpoint.

ezr-ondrej commented on internal/routes/all_routes.go:
```diff
@@ -23,15 +24,21 @@ func SetupRoutes(r *chi.Mux) {
 		r.Use(redocMiddleware)
-		r.Get("/openapi.json", api.ServeOpenAPISpec)
+		r.Route("/openapi.json", func(r chi.Router) {
+			r.Use(middleware.ETagMiddleware(15*time.Minute, "json-spec", api.ETagJSONSpec))
```
I've left the review here last time (sorry!)
I was trying to think of a way how we could introduce the middleware globally, but enable it per endpoint, but I guess this is the best way.

The only thing I've came up with, is to include the the tag in the return statement of the handler and middleware would be watching out for that, but it sounds bit ugly as we would change the Handler func definition just for that :thinking:

adiabramovitch replied:
So for now we are  caheching only in the openapi.json? and we are supposed to use this middleware in any endpoint which holds static information?

lzap replied:
Yeah this is the cleanest approach. Frankly, I don't see us using this for anything else than instance types and OpenAPI.

@adiabramovitch yes etag only works for content that does not change often (e.g. photos uploaded to a site). In our case instance types.

ETag is complementary to normal caching, it is not here to replace it at all. While it is possible to implement ETag with cache in tandem (could be timestamp of when some item was pulled from cache), I don't think it is worth the effort for our use case. If we cache something, let's just return it without ETag since it will be pretty fast anyway.

ezr-ondrej replied:
Well it will be fast, but the trafic still needs to happen, right, so if we are caching it, it's probably good idea to etag it as well IMHO. 200KB traffic is IMHO quite some safe

ezr-ondrej commented on internal/middleware/etag_middleware.go:
```diff
@@ -0,0 +1,39 @@
+			if etag == "" {
+				panic(InvalidETagErr)
+			}
+			etagValue := fmt.Sprintf("\"hms-%s-%s\"", name, etag)
```
Why do we want to have the `hms` prefix? HMS is internal name and only RH, so I'd rather not expose it (even in minor stuff user really don't see). Can we have `provisioning`/`prov` if we want prefix it?

lzap replied:
Sure, I was aiming for something short, it's not mandatory it's recommended to use a prefix to differentiate own content (we do have a name). Changed this to "pb".

ezr-ondrej commented on internal/middleware/etag_middleware.go:
```diff
@@ -0,0 +1,39 @@
+			logger := ctxval.Logger(r.Context()).With().Str("etag", etag).Str("etag_name", name).Logger()
+			logger.Trace().Msgf("Returned etag with Cache-Control '%s'", controlValue)
+
+			if match := r.Header.Get("If-None-Match"); match != "" {
```
I've got no idea how this is supposed to be implemented, so it's hard to review this part, why don't we use some external implementation?

adiabramovitch replied:
Yes, I am having the same issue here, can you explain in a few words?

lzap replied:
While there might be libraries providing such middleware, it is so trivial that I think it is simply worth having it in our codebase.

From Google: https://devcenter.heroku.com/articles/increasing-application-performance-with-http-cache-headers

ezr-ondrej reviewed (commented):
few comments, but more of general thoughts, feel free to disagree, I'd be ok to merge as is :)

lzap commented on api/embed.go:
```diff
@@ -3,12 +3,22 @@ package api
 var embeddedJSONSpec []byte
 
+// ETagJSONSpec contains sha sum of the SPEC json.
+var ETagJSONSpec string
```
Absolutely, now that I think about it, I refactored this code into a helper function in the middleware.

lzap replied:
Also added a test :)

lzap replied:
Ha forgot to unexport this variable, not needed anymore since there is a function now.

lzap commented:
Rebased.

ezr-ondrej commented on internal/middleware/etag_middleware.go:
```diff
@@ -0,0 +1,49 @@
+			logger.Trace().Msgf("Returned etag with Cache-Control '%s'", controlValue)
+
+			if match := r.Header.Get("If-None-Match"); match != "" {
+				if strings.Contains(match, etag) {
```
Shouldn't this be exact match? I know it's a minor thing, but I'm interested, why? Some whitespace issues?

lzap replied:
_The If-None-Match HTTP request header makes the request conditional. For [GET](https://developer.mozilla.org/en-US/docs/Web/HTTP/Methods/GET) and [HEAD](https://developer.mozilla.org/en-US/docs/Web/HTTP/Methods/HEAD) methods, the server will return the requested resource, with a [200](https://developer.mozilla.org/en-US/docs/Web/HTTP/Status/200) status, only if it doesn't have an [ETag](https://developer.mozilla.org/en-US/docs/Web/HTTP/Headers/ETag) matching the given ones. For other methods, the request will be processed only if the eventually existing resource's [ETag](https://developer.mozilla.org/en-US/docs/Web/HTTP/Headers/ETag) doesn't match any of the values listed._

From: https://developer.mozilla.org/en-US/docs/Web/HTTP/Headers/If-None-Match

ezr-ondrej replied:
ILT, thanks! (this is exactly what I meant by I don't fully understand how it should work, so it's hard to review) :)

ezr-ondrej reviewed (approved):
The helper is nice :+1: 
Do you think we should pull all the ETag constants into routes tho?
I feel kinda unease to have them scattered all around the code. Not a big deal tho :shrug: 
Apart of that and nit on the match comparison, I'm good here :+1:

lzap commented:
> I feel kinda unease to have them scattered all around the code.

Hmm, let me create a type with a function type to solve this. It won't be perfect since hash calculation will always be done somewhere else, but this will be vastly better.

Rebased, how about this?

lzap commented on internal/middleware/etag_middleware.go:
```diff
@@ -0,0 +1,67 @@
+	Value      string
+}
+
+type ETagValueFunc func() ETag
```
This is the function which returns the type above which contains all information for etag to be generated from.

ezr-ondrej reviewed (approved):
Nice interface! :orange_heart:
Great, I'm all up for it, we just need a rebase now.

lzap commented:
I rebased, I also added one more thing - duration of hash calculation is now logged via debug level. I expect it to be slower for Azure and other clouds, so I want to keep an eye on that to prevent slowing down of development cycles (app restarts). If this turns to be slow, let's make a setting for etag. Will be probably good tho.

ezr-ondrej reviewed (approved):
Thanks @lzap ! :+1:
